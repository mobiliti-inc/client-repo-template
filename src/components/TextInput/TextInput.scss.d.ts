// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'checkmark-label': string;
  'dark': string;
  'error': string;
  'error-message': string;
  'fades-in-out': string;
  'input-base': string;
  'input-date': string;
  'input-email': string;
  'input-file': string;
  'input-group': string;
  'input-image': string;
  'input-label': string;
  'input-number': string;
  'input-password': string;
  'input-tel': string;
  'input-text': string;
  'input-url': string;
  'password-visibility-indicator': string;
}
declare var cssExports: CssExports;
export = cssExports;
