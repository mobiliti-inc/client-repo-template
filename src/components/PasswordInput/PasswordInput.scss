@import "../../../styles/_variables.scss";

.input-validation-message-label{
   @include input-validation-text;
}


.input-container {
    display: block;
    width: 100%;
    margin-top: 14px;
  }

  .password-strength-label {
    display: flex;
    height: 14px;
    width: 47px;
    color: rgba(29, 49, 85, 0.6);
    font-family: $hanken-sans;
    font-size: 12px;
    line-height: 14px;
  }
  
  .password-weak,
  .password-medium,
  .password-strong {
    height: 14px;
    width: 47px;
    font-family: $hanken-sans;
    font-size: 12px;
    line-height: 14px;
    text-transform: capitalize;
    padding-left: 5%;
  }
  
  .password-strong {
    color: $green-check;
  }
  
  .password-medium {
    color: $orange;
  }
  
  .password-weak {
    color: $fast-car-red;
  }
  