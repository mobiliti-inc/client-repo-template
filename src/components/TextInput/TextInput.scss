@import "../../styles/_variables.scss";

.input-group {
  position: relative;
  display: flex;
  flex-direction: column;
}

.input-base {
  height: 50px;
  width: 100%;
  font-size: 22px;
  font-family: inherit;
  border: none;
  transition: all .5s;
  border-bottom: 1px solid rgba(29, 49, 85, 0.2);
  text-transform: capitalize;
  font-family: $hanken-sans;

  &:focus {
    outline: none;
    border-bottom: 3px solid $blue;
  }

  &::-webkit-input-placeholder {
    color: $darker-blue;
    opacity: 0.6;
  }

  &:placeholder-shown+.input-label {
    opacity: 0;
    visibility: hidden;
    transform: translateY(-40px);
  }

  &.error {
    border-bottom: 3px solid $fast-car-red;
  }
}

.input-email,
.input-password,
.input-number,
.input-image,
.input-text,
.input-file,
.input-date,
.input-url,
.input-tel {
  composes: input-base;
  text-transform: none;
}

.input-label {
  display: block;
  font-size: 12px;
  transform: translateY(-60px);
  color: $darker-blue;
  opacity: 0.6;
  font-family: $hanken-sans;
  line-height: 14px;
  text-transform: capitalize;
  transition: all .5s;
}

.checkmark-label {
  display: inline-block;
  position: absolute;
  top: 0px;
  right: 0px;
  height: 20px;
  width: 12px;
  border: solid $green-check;
  border-width: 0 2px 2px 0;
  cursor: pointer;
  transform: rotate(45deg);
}

.dark {
  color: $black;
  border-color: $black;

  &::placeholder {
    color: $black;
  }

  label {
    color: $black;
  }
}

.error-message {
  color: $fast-car-red;
  height: 14px;
  font-family: $hanken-sans;
  margin-top: 5px;
  font-size: 12px;
  line-height: 14px;
}

.password-visibility-indicator {
  display: inline-block;
  position: absolute;
  top: 15px;
  right: 0;
  line-height: 19px;
  color: $blue;
  font-family: $hanken-sans;
  font-size: 16px;
  border: none;
  background-color: transparent;
  text-transform: capitalize;
  cursor: pointer;

  &:active,
  &:focus {
    outline: none;
  }
}