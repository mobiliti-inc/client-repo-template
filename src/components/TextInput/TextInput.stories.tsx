import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import * as React from 'react';
import { TextInput } from '../';

storiesOf('TextInput', module)
	.add('default', () => (<div style={{ width: "300px" }}><TextInput type={text('type', 'text')}
		onChange={() => { }}
		id="text"
		value={text('value', '')}
		placeholder={text('value', '')}
		disabled={boolean('disabled', false)}
		isValid={boolean('isValid', false)}
		dark={boolean('dark', false)} /></div>
	)
	);