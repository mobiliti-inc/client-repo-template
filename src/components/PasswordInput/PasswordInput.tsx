import React from 'react';
import zxcvbn from 'zxcvbn';

import TextInput, { INPUT_TYPES } from '../TextInput/TextInput';
import styles from './PasswordInput.scss';

export const PASSWORD_STRENGTH_TYPE = {
	WEAK: 'weak',
	MEDIUM: 'medium',
	STRONG: 'strong'
};

export const ERROR_MESSAGES = {
	EMPTY_PASSWORD: 'Please enter a password.',
	PASSWORD_LESS_THAN_8: 'Please enter at least 8 characters.'
};

class PasswordInput extends React.Component {
	state = {
		value: '',
		passwordError: '',
		passwordStrength: PASSWORD_STRENGTH_TYPE.WEAK,
		isPasswordTextVisible: false,
	};

	inputChange = (event) => {
		const { value } = event.target;
		// Prevent empty value
		if (value !== value.trim()) {
			return event.preventDefault();
		}
		const passwordError = this.validatePassword(value);
		const passwordStrength = this.calculatePasswordStrength(value);

		this.setState({
			value,
			passwordError,
			passwordStrength
		});

		this.props.onChange({ password: value, passwordError, passwordStrength });
	}

	validatePassword = (value) => {
		if (!value.length) {
			return ERROR_MESSAGES.EMPTY_PASSWORD;
		} else if (value.length < 8) {
			return ERROR_MESSAGES.PASSWORD_LESS_THAN_8;
		}
		return '';
	}

	calculatePasswordStrength = (value) => {
		const results = zxcvbn(value);
		if (results.score <= 1) {
			return PASSWORD_STRENGTH_TYPE.WEAK;
		} else if (results.score === 2) {
			return PASSWORD_STRENGTH_TYPE.MEDIUM;
		}
		return PASSWORD_STRENGTH_TYPE.STRONG;
	}

	changePasswordVisibilityStatus = () => {
		this.setState(currentState => ({
			...currentState,
			isPasswordTextVisible: !currentState.isPasswordTextVisible
		}));
	}

	renderError = (error) => <label styleName="input-validation-message-label">{error}</label>

	render() {
		const { value, passwordError, isPasswordTextVisible, passwordStrength } = this.state;
		const { showPasswordStrength, id, inputError, placeholder, label, name } = this.props;
		return (
			<div>
				<TextInput
					{...this.props}
					type={isPasswordTextVisible ? INPUT_TYPES.TEXT : INPUT_TYPES.PASSWORD}
					id={id}
					placeholder={placeholder || 'Password'}
					label={label || 'password'}
					name={name || 'password'}
					isValid={!passwordError.length}
					onChange={this.inputChange}
					value={value}
					shouldShowVisibility
					changePasswordVisibilityStatus={this.changePasswordVisibilityStatus}
				/>

				{shouldShowVisibility &&
						value &&
						<button
							onClick={changePasswordVisibilityStatus}
							styleName="password-visibility-indicator"
						>
							{INPUT_TYPES.PASSWORD === type ?
								'Show' :
								'Hide'}
						</button>
				}

				{ passwordError.length > 0 && this.renderError(passwordError)}

				{ inputError && inputError.length > 0 && this.renderError(inputError)}

				{
					inputError.length === 0 &&
					passwordError.length === 0 &&
					value.length > 0 &&
					showPasswordStrength &&
					<label styleName="password-strength-label">
						Strength:<span styleName={`password-${passwordStrength}`}>{passwordStrength}</span>
					</label>
				}
			</div>);
	}
}

PasswordInput.propTypes = {
	onChange: PropTypes.func.isRequired,
	showPasswordStrength: PropTypes.bool,
	id: PropTypes.string,
	inputError: PropTypes.string,
	placeholder: PropTypes.string,
	label: PropTypes.string,
	name: PropTypes.string,
};

PasswordInput.defaultProps = {
	showPasswordStrength: true,
	id: 'password',
	inputError: '',
	label: null,
	placeholder: null,
	name: null,
};

export default CSSModules(PasswordInput, styles, { allowMultiple: true });
