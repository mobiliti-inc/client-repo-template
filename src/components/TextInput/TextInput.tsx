import * as React from 'react';
import cx from 'classnames';

import './TextInput.scss';

export const INPUT_TYPES = {
	PASSWORD: 'password',
	NUMBER: 'number',
	EMAIL: 'email',
	IMAGE: 'image',
	TEXT: 'text',
	FILE: 'file',
	DATE: 'date',
	PHONE: 'tel',
	URL: 'url'
};

interface TextInputProps {
	type: string;
	onChange: (...args: any[]) => any;
	value: string;
	id: string;
	placeholder?: string;
	disabled?: boolean;
	isValid: boolean;
	inputDidChange?: boolean;
	onBlur?: (...args: any[]) => any;
	label?: string;
	name?: string;
	dark?: boolean;
	className?: string;
	errorMessage?: React.ReactNode | string;
	shouldShowVisibility?: boolean;
	changePasswordVisibilityStatus?: (...args: any[]) => any;
}

const TextInput: React.FC<TextInputProps> = (props) => {
	const {
		placeholder,
		type,
		onChange,
		disabled,
		label,
		value,
		name,
		id,
		isValid,
		dark,
		className,
		inputDidChange,
		onBlur,
		errorMessage
	} = props;

	const [inputValue, setInputValue] = React.useState('');

	React.useEffect(() => {
		setInputValue(value);
	}, [value]);

	const baseStyles = cx(`input-${type}`, {
		error: !isValid
	});

	const styleName = cx(baseStyles, { dark }, className);

	// TODO: We really shouldn't use any here ðŸ˜“ but rather React.FormEvent<HTMLInputElement>, value was kinda nagging us
	const handleChange = (event: any) => {
		setInputValue(event.target.value);
		onChange(event);
	};

	return (
		<React.Fragment>
			<div styleName="input-group">
				<input
					placeholder={placeholder}
					styleName={styleName}
					className={className}
					disabled={disabled}
					onChange={(event) => handleChange(event)}
					type={type}
					value={inputValue}
					name={name}
					id={id}
					onBlur={onBlur}
				/>
				{label && (
					<label styleName="input-label" htmlFor={id}>
						{label}
					</label>
				)}
				{isValid && inputDidChange && value && value.length > 0 && <label styleName="checkmark-label" />}
				{!isValid && errorMessage && <span styleName="error-message">{errorMessage}</span>}
			</div>
		</React.Fragment>
	);
};

export default TextInput;
