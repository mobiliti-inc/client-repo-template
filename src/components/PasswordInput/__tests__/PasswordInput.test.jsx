import PasswordInput, { PASSWORD_STRENGTH_TYPE, ERROR_MESSAGES } from '../PasswordInput';
import TextInput from '../../../components/TextInput/TextInput';

describe('PasswordInput', () => {
	const initialProps = {
		onChange: jest.fn()
	};
	it('renders', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		matchSnapshot(tree);
	});
	it('renders one <TextInput /> components', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.find(TextInput)).toHaveLength(1);
	});

	it('updates state correctly on input change', () => {
		const event = { target: { value: '1234test' } };
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		tree.instance().inputChange(event);
		expect(tree.instance().state).toEqual({
			value: '1234test',
			passwordError: '',
			passwordStrength: PASSWORD_STRENGTH_TYPE.WEAK,
			isPasswordTextVisible: false,
		});
	});

	it('should not updates state on input change when value contains whitespace', () => {
		const event = { target: { value: '1234test ' }, preventDefault: () => { } };
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		tree.instance().inputChange(event);
		expect(tree.instance().state.value).not.toEqual(event.target.value);
	});

	it('updates password visibility status correctly when changePasswordVisibilityStatus method is called', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().state.isPasswordTextVisible).toEqual(false);
		tree.instance().changePasswordVisibilityStatus();
		expect(tree.instance().state.isPasswordTextVisible).toEqual(true);
	});

	it('should return please enter a password if password value is empty', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().validatePassword('')).toEqual(ERROR_MESSAGES.EMPTY_PASSWORD);
	});

	it('should return Please enter at least 8 characters if password length is less than 8', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().validatePassword('1234')).toEqual(ERROR_MESSAGES.PASSWORD_LESS_THAN_8);
	});

	it('should return a weak password strength when calculatePasswordStrength method is called with a weak password value', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().calculatePasswordStrength('12345678')).toEqual(PASSWORD_STRENGTH_TYPE.WEAK);
	});

	it('should return a medium password strength when calculatePasswordStrength method is called with the given password below', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().calculatePasswordStrength('aaafcfjh')).toEqual(PASSWORD_STRENGTH_TYPE.MEDIUM);
	});

	it('should return strong password strength when calculatePasswordStrength method is called with the given password below', () => {
		const tree = shallow(
			<PasswordInput {...initialProps} />);
		expect(tree.instance().calculatePasswordStrength('aaaf%^&**Gcfjh')).toEqual(PASSWORD_STRENGTH_TYPE.STRONG);
	});
});
