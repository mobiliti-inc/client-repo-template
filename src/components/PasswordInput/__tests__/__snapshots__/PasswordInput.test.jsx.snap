// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput renders 1`] = `
<div>
  <TextInput
    changePasswordVisibilityStatus={[Function]}
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    id="password"
    inputDidChange={false}
    inputError=""
    isValid={true}
    label="password"
    name="password"
    onBlur={null}
    onChange={[Function]}
    placeholder="Password"
    shouldShowVisibility={true}
    showPasswordStrength={true}
    type="password"
    value=""
  />
</div>
`;
